src/mod_ping.erl

```diff
@@ -1,226 +1,166 @@
 
 -module(mod_ping).
--author('bjc@kublai.com').
+-author('piotr.nosek@erlang-solutions.com').
 
--behaviour(gen_mod).
--behaviour(gen_server).
--behaviour(mongoose_module_metrics).
+-behavior(gen_mod).
+-behavior(ejabberd_c2s_info_handler).
 -xep([{xep, 199}, {version, "2.0"}]).
 -include("mongoose.hrl").
 -include("jlib.hrl").
 
--define(SUPERVISOR, ejabberd_sup).
 -define(DEFAULT_SEND_PINGS, false). % bool()
 -define(DEFAULT_PING_INTERVAL, 60). % seconds
 -define(DEFAULT_PING_REQ_TIMEOUT, 32).
 
-%% API
--export([start_link/2, start_ping/2, stop_ping/2]).
+%% C2S custom info handler
+-export([handle_c2s_info/3]).
 
 %% gen_mod callbacks
 -export([start/2, stop/1]).
 
-%% gen_server callbacks
--export([init/1, terminate/2, handle_call/3, handle_cast/2,
-         handle_info/2, code_change/3]).
-
 %% Hook callbacks
 -export([iq_ping/4,
          user_online/4,
          user_offline/5,
          user_send/4,
-         user_keep_alive/2,
-         user_ping_response/4]).
-
--record(state, {host = <<"">>,
-                send_pings = ?DEFAULT_SEND_PINGS,
-                ping_interval = ?DEFAULT_PING_INTERVAL,
-                timeout_action = none,
-                ping_req_timeout = ?DEFAULT_PING_REQ_TIMEOUT,
-                timers = maps:new()}).
-
--type state() :: #state{}.
+         user_ping_response/4,
+         user_keep_alive/2]).
 
 %%====================================================================
-%% API
+%% Info Handler
 %%====================================================================
-start_link(Host, Opts) ->
-    Proc = gen_mod:get_module_proc(Host, ?MODULE),
-    gen_server:start_link({local, Proc}, ?MODULE, [Host, Opts], []).
 
-start_ping(Host, JID) when JID#jid.lresource =/= <<>> ->
-    %% Guard check above ensures we are not adding a user without a
-    %% resource by accident. If this happens the server ends up in a
-    %% hot loop of iq-error replies.
-    Proc = gen_mod:get_module_proc(Host, ?MODULE),
-    gen_server:cast(Proc, {start_ping, JID});
-start_ping(_Host, _JID) -> ok.
+handle_c2s_info(init, HandlerState, Params) ->
+    start_ping_timer(HandlerState, Params);
+handle_c2s_info(send_ping, HandlerState, #{jid := JID, server := Server} = Params) ->
+    route_ping_iq(JID, Server),
+    start_ping_timer(HandlerState, Params);
+handle_c2s_info(timeout, HandlerState, #{jid := JID, server := Server}) ->
+    mongoose_hooks:user_ping_timeout(Server, ok, JID),
+    case gen_mod:get_module_opt(Server, ?MODULE, timeout_action, none) of
+        kill -> ejabberd_c2s:stop(self());
+        _ -> ok
+    end,
+    HandlerState.
+
+start_ping_timer(HandlerState, #{server := Server}) ->
+    start_ping_timer(HandlerState, Server);
+start_ping_timer(HandlerState, Server) ->
+    cancel_timer(HandlerState),
+    PingInterval = gen_mod:get_module_opt(Server, ?MODULE, ping_interval, ?DEFAULT_PING_INTERVAL),
+    {ok, Tref} = timer:apply_after(PingInterval * 1000,
```

Comment:
Uff... I'm way less convinced about that one. That goes through a `gen_server` called `timer_server`, hence it might again route all timeouts through a single process 🙁

Reply:
Yes, usage of the `timer` module is discouraged by the [Erlang Efficiency Guide](http://erlang.org/doc/efficiency_guide/commoncaveats.html#timer-module)